src/granite/associations.cr

```diff
@@ -1,75 +1,135 @@
 module Granite::Associations
   macro belongs_to(model, **options)
-    {% if model.is_a? TypeDeclaration %}
-      {% method_name = model.var %}
-      {% class_name = model.type %}
-    {% else %}
-      {% method_name = model.id %}
-      {% class_name = options[:class_name] || model.id.camelcase %}
-    {% end %}
+    {%
+      nilable = false
+      if model.is_a? TypeDeclaration
+        name = model.var.id
+        type = model.type
+      else
+        name = model.id
+        type = options[:class_name] || name.camelcase
+      end
+      if type.is_a? Union
+        nilable = type.types.any?(&.resolve?.== Nil)
+        type = type.types.find(&.resolve?.!= Nil)
+      end
+      if nilable
+        nilable_method = name
+        not_nil_method = name + "!"
+      else
+        nilable_method = name + "?"
+        not_nil_method = name
+      end
+      type = type.id
+    %}
 
     {% if options[:foreign_key] && options[:foreign_key].is_a? TypeDeclaration %}
-      {% foreign_key = options[:foreign_key].var %}
+      {% foreign_key = options[:foreign_key].var.id %}
       column {{options[:foreign_key]}}
     {% else %}
-      {% foreign_key = method_name + "_id" %}
+      {% foreign_key = name + "_id" %}
       column {{foreign_key}} : Int64?
     {% end %}
-    {% primary_key = options[:primary_key] || "id" %}
+    {% primary_key = (options[:primary_key] || "id").id %}
 
-    def {{method_name.id}} : {{class_name.id}}?
-      if parent = {{class_name.id}}.find_by({{primary_key.id}}: {{foreign_key.id}})
-        parent
-      else
-        {{class_name.id}}.new
-      end
+    @[JSON::Field(ignore: true)]
+    @[YAML::Field(ignore: true)]
+    @{{name}} : {{type}}?
```

Comment:
I'm not sure how I feel about the caching side of this yet.  I don't like that the ivar is hidden away and can't be altered, via annotations for example.  I'm also not sure about this being the default behavior.  I thought about caching a while ago and my thought was the query itself should be cached, so that an extra ivar doesn't have to be added to the model.

Reply:
@Blacksmoke16 Another purpose of this pr is to solve the N + 1 problem, which cannot be solved by caching the query.
There may be another better approach, but the best way I came up with is this.

Reply:
> There may be another better approach, but the best way I came up with is this.

I'd rather hold off on this change and discuss alternative solutions.  In its current state this PR would not play well with any third party serializer.  If anything it should be an opt in feature via a module or something.

Reply:
@Blacksmoke16 I see your concern regarding serialization and agree that we should find another alternative approach.

@c910335 I am wondering if we need the parent query to include the children records.  Would that solve the N+1 and allow the db to perform caching?

Reply:
@drujensen 
Something like this.

```sql
SELECT * FROM `children` WHERE `children`.`parent_id` IN ('2', '3', '5', '7', '11');
```

Then assign each child to the instance variable of its parent in Crystal.